test/1_stdlib/Tuple.swift.gyb

```diff
@@ -0,0 +1,172 @@
+
+% maxArity = 6 # the highest arity the operators are defined for
+
+func testEquality<A : Equatable, B : Equatable, C : Equatable>(
+  lhs: (A,B,C), equal: Bool, to rhs: (A,B,C),
+  //===--- TRACE boilerplate ----------------------------------------------===//
+  @autoclosure _ message: ()->String = "",
+  showFrame: Bool = true,
+  stackTrace: SourceLocStack = SourceLocStack(),  
+  file: String = __FILE__, line: UInt = __LINE__
+) {
+  let trace = stackTrace.pushIf(showFrame, file: file, line: line)
+  expectEqual(equal, lhs == rhs, stackTrace: trace)
+  expectEqual(equal, rhs == lhs, stackTrace: trace)
+  expectEqual(!equal, lhs != rhs, stackTrace: trace)
+  expectEqual(!equal, rhs != lhs, stackTrace: trace)
+}
+
+TupleTestSuite.test("Tuple/equality") {
+  testEquality((1,2,3), equal: true, to: (1,2,3))
+  testEquality((1,2,3), equal: false, to: (1,2,4))
+  testEquality((1,2,3), equal: false, to: (1,3,3))
+  testEquality((1,2,3), equal: false, to: (2,2,3))
+  testEquality((1,"2",3), equal: true, to: (1,"2",3))
+  testEquality((1,"2",3), equal: false, to: (1,"3",3))
+  testEquality(("one", 2.2, 3..<5), equal: true, to: ("one", 2.2, 3..<5))
+
+  testEquality((1.0, 2.0, 3.0), equal: false, to: (1.0, 2.0, .NaN))
+  testEquality((1.0, 2.0, 3.0), equal: false, to: (1.0, .NaN, 3.0))
+  testEquality((1.0, 2.0, 3.0), equal: false, to: (.NaN, 2.0, 3.0))
+  testEquality((1.0, 2.0, 3.0), equal: false, to: (.NaN, .NaN, .NaN))
+  testEquality((1.0, 2.0, Float.NaN), equal: false, to: (1.0, 2.0, 3.0))
+  testEquality((1.0, 2.0, Float.NaN), equal: false, to: (1.0, 2.0, Float.NaN))
+  testEquality((Float.NaN, Float.NaN, Float.NaN), equal: false, to: (.NaN, .NaN, .NaN))
+  testEquality((Float.NaN, Float.NaN, Float.NaN), equal: false, to: (1.0, 2.0, 3.0))
+
+  expectTrue((1,2) == (1,2))
+  expectTrue((1,2) != (1,3))
+  expectTrue((1,2,3,4) == (1,2,3,4))
+  expectTrue((1,2,3,4) != (1,2,3,3))
+  expectTrue((1,2,3,4,5) == (1,2,3,4,5))
+  expectTrue((1,2,3,4,5) != (1,2,3,4,4))
+  expectTrue((1,2,3,4,5,6) == (1,2,3,4,5,6))
+  expectTrue((1,2,3,4,5,6) != (1,2,3,4,5,5))
+}
+
+TupleTestSuite.test("Tuple/equality/sanity-check") {
+  // sanity check all arities
+% for arity in range(2,maxArity+1):
+%   a = str(tuple(range(1, arity+1)))
+%   b = "({}, 0)".format(", ".join([str(i) for i in range(1,arity)]))
+%   c = "(0, {})".format(", ".join([str(i) for i in range(2,arity+1)]))
+  expectTrue(${a} == ${a})
+  expectTrue(${a} != ${b})
+  expectTrue(${b} != ${a})
+  expectTrue(${a} != ${c})
+  expectTrue(${c} != ${a})
+% end
+}
+
+enum Ordering : Equatable {
+  case LessThan
+  case EqualTo
+  case GreaterThan
+  case UnorderedWith // Comparable defines strict total order, but Float disobeys that with NaN
+
+  func flip() -> Ordering {
+    switch self {
+    case .LessThan: return .GreaterThan
+    case .EqualTo: return .EqualTo
+    case .GreaterThan: return .LessThan
+    case .UnorderedWith: return .UnorderedWith
+    }
+  }
```

Comment:
You're right. I created it because I was using existing floating-point tests as inspiration, but didn't end up using it because there was no reason to do so.